### ISSUE TITLE
在Tensor中无法使用Tensor.bool类型进行布尔索引

### ISSUE THREAD

**Issue by zhude233** [PaddlePaddle/Paddle#34343]
PaddlePaddle版本：2.1

如题所示
![WOIUO_IG5HAWNX_864LD9S2](https://user-images.githubusercontent.com/60126992/126747488-8825e051-108c-480d-864e-22432950f221.png)

**Comment by zhude233** [comment]
如何在tensor下进行布尔索引

**Comment by visualDust** [comment]
do consider using `paddle.mask_select`.
example:
```python
import paddle

x = paddle.to_tensor([[1.0, 2.0, 3.0, 4.0],
                      [5.0, 6.0, 7.0, 8.0],
                      [9.0, 10.0, 11.0, 12.0]])
mask = paddle.to_tensor([[True, False, False, False],
                         [True, True, False, False],
                         [True, False, False, False]])
out = paddle.masked_select(x, mask)
#[1.0 5.0 6.0 9.0]
```
doc at https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/masked_select_en.html

**Comment by lytning98** [comment]
> do consider using `paddle.mask_select`.
> example:
> 
> ```python
> import paddle
> 
> x = paddle.to_tensor([[1.0, 2.0, 3.0, 4.0],
>                       [5.0, 6.0, 7.0, 8.0],
>                       [9.0, 10.0, 11.0, 12.0]])
> mask = paddle.to_tensor([[True, False, False, False],
>                          [True, True, False, False],
>                          [True, False, False, False]])
> out = paddle.masked_select(x, mask)
> #[1.0 5.0 6.0 9.0]
> ```
> 
> doc at https://www.paddlepaddle.org.cn/documentation/docs/en/api/paddle/masked_select_en.html

What if I want to combine mask indexing with assignment, i.e. `masked_fill` a tensor like what in PyTorch?
E.g., how to implement a numpy-or-pytorch style statement `x[mask]=0.0` with Paddle?

**Comment by ceci3** [comment]
@lytning98 可以尝试用paddle.where 来实现masked_fill 功能哈

**Comment by visualDust** [comment]
@lytning98
Boolean array indexing is a feature from NumPy. indexing with a list of booleans performs masking. I can't figure out how is indexing with a boolean array is implemented in PyTorch using the indexing operator, but many issues under the PyTorch repo showed there must be some difference from NumPy. 
In Tensorflow we have `boolean_mask` and `masked_select` in PyTorch. I haven't seen another framework except PyTorch provide the feature that supports masking using an indexing operator. additionally. Apparently using functions like `mask_select` is a common way among these frameworks. Currently, DO CONSIDER using `mask_select` unless the framework has a stable implementation that allows you to use the indexing operator. (I don't know if it's correct.)
